README.md
# Diskuss

[![Build Status](https://travis-ci.org/SteeveDroz/diskuss.svg?branch=master)](https://travis-ci.org/SteeveDroz/diskuss)

Diskuss is a Node.js chat server that offers an API but no official client.

## Install

1. This application requires **[node.js](https://nodejs.org/en/download/)** and **NPM** (included with Node.js), please install it if you haven't.
2. Download the [latest version](https://github.com/SteeveDroz/diskuss/releases/latest) of diskuss and unzip it.
3. From the directory where `server.js` is located, use the command line `npm install` to download and install dependencies.
4. Run the server with `npm start`.
5. Open a browser (or IE if you don't have any) and connect to your server on port 8081\. If you access that address from the same computer as the one you started the server on, simply follow **[this link](http://127.0.0.1:8081)**. The description of the API is located at the root of the application.

## About Diskuss

### The project

Diskuss was originally created for a school project. It has been made by a teacher whose students had to learn about RESTful APIs. The goal of the exercise was to create a client for Diskuss, each student using a different language and implementing what they thought was interresting to add client-side.

As many other chat servers and clients exist on the market, there are no practical uses for such a half-made network, except the one described hereinabove. If you are a teacher of any kind and you want to use any of my work with your students, please _do_ and give me a feedback if you have the time.

### Release notes

Release notes can be found in the **[change log](changelog.md)**.


changelog.md
# Changelog

Here are all the updates since v1.0

## [dev]: Current updates

None, `dev` is up to date.

## [v1.6.0]: Kicking update
The change log is more taken care of:

- Indenting has been fixed.
- The "dev" chapter has been added.

Users can now be kicked out of channels by owner.
Users can also be banned and unbanned by owner.
Users connected to the server all receive a notice when a new channel is created.

## [v1.5.0]: Information update

Informations about channels can be obtained.

Regression bugs in the unit tests made the build fail, they have been fixed.

## [v1.4.2]: Moderated update

Fixing a bug where two people wanting the same username crashed the server.

Fixing a bug where the disconnecting of idling users wasn't notified to other users.

## [v1.4.1]: Moderated update

Server now allows [CORS](https://en.wikipedia.org/wiki/Cross-origin_resource_sharing), useful for AJAX developpers.

## [v1.4.0]: Moderated update

Channels now belong to the user who first joined it.

Changes:

- An owner (and only them) can:

  - Change the description of the channel
  - Keep or release the channel
  - Permanently transfer their ownership of the channel to another user

- Update of all the notices example to include owners
- Various bug fixes

## [v1.3.0]: Channel update

**:warning: This update breaks a little backwards compatibility! The notices have changed**

The "channel" side of the application has been developped.

- Channel descriptions can be changed.
- Channels can be kept even if no user is in (the description will remain, more features are to be added)
- Notices have been changed. When a channel name was part of it, it has been replaced by a channel object containing all the useful data.
- Use of JavaScript in the API to ease the writing of new features.
- Various bug fixes.

## [v1.2.0]: Logical update

**:warning: This update breaks backwards compatibility!**

To be more logical, the responses have been changed. It was, for example, not logical at all that leaving a channel would return the user object of oneself.

The following responses have been updated:

- **Disconnection** now sends a JSON object with `status` property.
- **Join channel** now sends a JSON object containing a `channel` and a `users` propoerties.
- **Talk in channel** now sends a JSON object containing a `status` and a `message` properties.
- **Leave channel** now sends a JSON object containing a `status` and a `channel` properties.
- **Send private message** now sends a JSON object containing a `status`, a `message` and a `recipient` properties.

## [v1.1.2]: Idle update

The following features have been fixed:

- The last user leaving a channel destroys it.
- Asking for a list of channels returns the expected value.

## [v1.1.1]: Idle update

The informations about v1.1 have been added to the API.

## [v1.1.0]: Idle update

Diskuss uses notices to check if users are still connected. The server postulates that any client that hasn't checked for notices in the last 5 seconds must be disconnected and removes it from the list of users.

## [v1.0.0]: Original release

The application was created from scratch, the current version offers the following features:

- Display the API
- Connect to a server
- Disconnect from the server
- Request server informations
- List connected users
- List channels
- Ask informations about a user (whois)
- Join a channel
- Talk in a channel
- Leave a channel
- Send a private message to another user
- Fetch notices

[dev]: https://github.com/SteeveDroz/diskuss/compare/master...dev
[v1.0.0]: https://github.com/SteeveDroz/diskuss/compare/v0.1-alpha...v1.0.0
[v1.1.0]: https://github.com/SteeveDroz/diskuss/compare/v1.0.0...v1.1.0
[v1.1.1]: https://github.com/SteeveDroz/diskuss/compare/v1.1.0...v1.1.1
[v1.1.2]: https://github.com/SteeveDroz/diskuss/compare/v1.1.1...v1.1.2
[v1.2.0]: https://github.com/SteeveDroz/diskuss/compare/v1.1.2...v1.2.0
[v1.3.0]: https://github.com/SteeveDroz/diskuss/compare/v1.2.0...v1.3.0
[v1.4.0]: https://github.com/SteeveDroz/diskuss/compare/v1.3.0...v1.4.0
[v1.4.1]: https://github.com/SteeveDroz/diskuss/compare/v1.4.0...v1.4.1
[v1.4.2]: https://github.com/SteeveDroz/diskuss/compare/v1.4.1...v1.4.2
[v1.5.0]: https://github.com/SteeveDroz/diskuss/compare/v1.4.2...v1.5.0
[v1.6.0]: https://github.com/SteeveDroz/diskuss/compare/v1.5.0...v1.6.0


spec/UserSpec.js
var User = require('../app/models/User')

describe("User", function() {
    it("has nick, id and channels", function() {
        let user = new User("Yoan")
        expect(user.nick).toEqual("Yoan")
        expect(user.id).not.toBeUndefined()
        expect(user.channels).not.toBeUndefined()
    })

    it("must have a nick or be anonymous", function() {
        let user = new User()
        expect(user.nick).toEqual("Anonymous")
    })
})


spec/MultiSpec.js
const app = require('../server'),
    request = require('supertest')

describe('Multiuser app', function() {
	const agent = request.agent(app)
	
	let id1, id2, id3
	
	it('connects the first user', function(done) {
		agent.post('/users/register/user1/')
			.end(function(err, res) {
				expect(200)
				id1 = res.body.id
				expect(id1).not.toBeUndefined()
				done()
			})
	})
	
	it('connects the second user', function(done) {
		agent.post('/users/register/user2/')
			.end(function(err, res) {
				expect(200)
				id2 = res.body.id
				expect(id2).not.toBeUndefined()
				done()
			})
	})
	
	it('asks for the list of users', function(done) {
		agent.get('/users/')
			.end(function(err, res) {
				expect(200)
				const users = res.body
				expect(users).not.toBeUndefined()
				if (users !== undefined) {
					expect(users.length).toBe(2)
					if (users.length == 2) {
						expect(users[0].nick).toBe('user1')
						expect(users[1].nick).toBe('user2')
					}
				}
				done()
			})
	})
    
    it('asks for the list of channels before joining', function(done) {
        agent.get('/channels/')
            .end(function(err, res) {
                expect(200)
                const channels = res.body
                expect(channels).not.toBeUndefined()
                if (channels !== undefined) {
                    expect(channels.length).toBe(0)
                }
                done()
            })
    })
	
	it('makes user1 enter a channel', function(done) {
		agent.put('/user/' + id1 + '/channels/talk/join/')
			.end(function(err, res) {
				expect(200)
				const users = res.body
				expect(users).not.toBeUndefined()
				if (users !== undefined) {
					expect(users.length).toBe(1)
					if (users.length == 1) {
						expect(users[0].nick).toBe('user1')
					}
				}
				done()
			})
	})
	
    it('asks for the list of channels after joining', function(done) {
        agent.get('/channels/')
            .end(function(err, res) {
                expect(200)
                const channels = res.body
                expect(channels).not.toBeUndefined()
                if (channels !== undefined) {
                    expect(channels.length).toBe(1)
                    if (channels.length == 1) {
                        expect(channels[0]).toBe('talk')
                    }
                }
                done()
            })
    })
	
	it('makes user1 enter another channel', function(done) {
		agent.put('/user/' + id1 + '/channels/chat/join/')
			.end(function(err, res) {
				expect(200)
				const users = res.body
				expect(users).not.toBeUndefined()
				if (users !== undefined) {
					expect(users.length).toBe(1)
					if (users.length == 1) {
						expect(users[0].nick).toBe('user1')
					}
				}
				done()
			})
	})
	
    it('asks for the list of channels after joining a second one', function(done) {
        agent.get('/channels/')
            .end(function(err, res) {
                expect(200)
                const channels = res.body
                expect(channels).not.toBeUndefined()
                if (channels !== undefined) {
                    expect(channels.length).toBe(2)
                    if (channels.length == 2) {
                        expect(channels[0]).toBe('talk')
                        expect(channels[1]).toBe('chat')
                    }
                }
                done()
            })
    })
	
	it('makes user1 talk in channel', function(done) {
		agent.put('/user/' + id1 + '/channels/talk/say')
			.send({message: 'Message 1'})
			.end(function(err, res) {
				expect(200)
				const user = res.body
				expect(user).not.toBeUndefined()
				if (user !== undefined) {
					expect(user.nick).toBe('user1')
				}
				done()
			})
	})
	
	it('makes user2 enter a channel', function(done) {
		agent.put('/user/' + id2 + '/channels/talk/join/')
			.end(function(err, res) {
				expect(200)
				const users = res.body
				expect(users).not.toBeUndefined()
				if (users !== undefined) {
					expect(users.length).toBe(2)
					if (users.length == 2) {
						expect(users[0].nick).toBe('user1')
						expect(users[1].nick).toBe('user2')
					}
				}
				done()
			})
	})
	
	it('makes user1 talk again in channel', function(done) {
		agent.put('/user/' + id1 + '/channels/talk/say/')
			.send({message: 'Message 2'})
			.end(function(err, res) {
				expect(200)
				const user = res.body
				expect(user).not.toBeUndefined()
				if (user !== undefined) {
					expect(user.nick).toBe('user1')
				}
				done()
			})
	})
    
    it('makes user1 send a private message to user2', function(done) {
        agent.put('/user/' + id1 + '/message/user2/')
            .send({ message: 'Private message' })
            .end(function(err, res) {
                expect(200)
                const user = res.body
                expect(user).not.toBeUndefined()
                expect(user.nick).toBe('user2')
                done()
            })
    })

	
	it('makes user1 check for notices', function(done) {
		agent.get('/user/' + id1 + '/notices/')
			.end(function(err, res) {
                expect(200)
                const notices = res.body
                expect(notices).not.toBeUndefined()
                expect(notices.length).toEqual(5)
				if (notices.length == 5) {
                    expect(notices[0].type).toEqual('channelJoin')
                    expect(notices[0].nick).toEqual('user1')
                    expect(notices[0].channel).toEqual('talk')
                    expect(notices[0].time).not.toBeUndefined()
                    if (notices[0].time !== undefined) {
                        expect(notices[0].time.length).toBe(24)
                    }

                    expect(notices[1].type).toEqual('channelJoin')
                    expect(notices[1].nick).toEqual('user1')
                    expect(notices[1].channel).toEqual('chat')
                    expect(notices[1].time).not.toBeUndefined()
                    if (notices[1].time !== undefined) {
                        expect(notices[1].time.length).toBe(24)
                    }

                    expect(notices[2].type).toEqual('channelMessage')
                    expect(notices[2].nick).toEqual('user1')
                    expect(notices[2].channel).toEqual('talk')
                    expect(notices[2].message).toEqual('Message 1')
                    expect(notices[2].time).not.toBeUndefined()
                    if (notices[2].time !== undefined) {
                        expect(notices[2].time.length).toBe(24)
                    }

                    expect(notices[3].type).toEqual('channelJoin')
                    expect(notices[3].nick).toEqual('user2')
                    expect(notices[3].channel).toEqual('talk')
                    expect(notices[3].time).not.toBeUndefined()
                    if (notices[3].time !== undefined) {
                        expect(notices[3].time.length).toBe(24)
                    }

                    expect(notices[4].type).toEqual('channelMessage')
                    expect(notices[4].nick).toEqual('user1')
                    expect(notices[4].channel).toEqual('talk')
                    expect(notices[4].message).toEqual('Message 2')
                    expect(notices[4].time).not.toBeUndefined()
                    if (notices[4].time !== undefined) {
                        expect(notices[4].time.length).toBe(24)
                    }
				}
				done()
			})
	})
	
	it('makes user2 check for notices', function(done) {
		agent.get('/user/' + id2 + '/notices/')
			.end(function(err, res) {
                expect(200)
                const notices = res.body
                expect(notices).not.toBeUndefined()
                expect(notices.length).toEqual(3)
				if (notices.length == 3) {
                    expect(notices[0].type).toEqual('channelJoin')
                    expect(notices[0].nick).toEqual('user2')
                    expect(notices[0].channel).toEqual('talk')
                    expect(notices[0].time).not.toBeUndefined()
                    if (notices[0].time !== undefined) {
                        expect(notices[0].time.length).toBe(24)
                    }

                    expect(notices[1].type).toEqual('channelMessage')
                    expect(notices[1].nick).toEqual('user1')
                    expect(notices[1].channel).toEqual('talk')
                    expect(notices[1].message).toEqual('Message 2')
                    expect(notices[1].time).not.toBeUndefined()
                    if (notices[1].time !== undefined) {
                        expect(notices[1].time.length).toBe(24)
                    }

                    expect(notices[2].type).toBe('privateMessage')
                    expect(notices[2].sender).toBe('user1')
                    expect(notices[2].message).toBe('Private message')
                    expect(notices[2].time).not.toBeUndefined()
                    if (notices[2].time !== undefined) {
                        expect(notices[2].time.length).toBe(24)
                    }
				}
				done()
			})
	})
	
	it('disconnects user1', function(done) {
		agent.del('/user/' + id1 + '/disconnect/')
			.end(function(err, res) {
				expect(200)
				const version = res.body
				expect(version).not.toBeUndefined()
				done()
			})
	})
    	
	it('makes user2 check for notices again', function(done) {
		agent.get('/user/' + id2 + '/notices/')
			.end(function(err, res) {
                expect(200)
                const notices = res.body
                expect(notices).not.toBeUndefined()
                expect(notices.length).toEqual(1)
				if (notices.length == 1) {
					expect(notices[0].type).toEqual('channelLeave')
					expect(notices[0].nick).toEqual('user1')
					expect(notices[0].channel).toEqual('talk')
                    expect(notices[0].time).not.toBeUndefined()
                    if (notices[0].time !== undefined) {
                        expect(notices[0].time.length).toBe(24)
                    }
				}
            done()
			})
	})
    
    it('connects user3', function(done) {
        agent.post('/users/register/user3/')
            .end(function(err, res) {
                expect(200)
                id3 = res.body.id
                expect(id3).not.toBeUndefined()
                done()
            })
    })
    
    it('waits for 6 seconds', function(done) {
        setTimeout(done, 6000 /* 6 seconds */)
    }, 10000 /* Jasmine timeout */)
    
    it('makes user2 check for notices after 6 seconds', function(done) {
        agent.get('/user/' + id2 + '/notices/')
            .end(function(err, res) {
                expect(200)
                done()
            })
    })
    
    it('makes user3 check for notices', function(done) {
        agent.get('/user/' + id3 + '/notices/')
            .end(function(err, res) {
                expect(404)
                const notice = res.body
                expect(notice).not.toBeUndefined()
                if (notice !== undefined) {
                    expect(notice.error).toBe('Unknown user ID')
                }
                done()
            })
    })
})


app/Store.js
class Store {
    constructor() {
        this._users = {}
        this._channels = {}
    }

    get users() {
        return this._users
    }

    get channels() {
        return this._channels
    }

    getUser(id) {
        return this.users[id]
    }

    getUserByNick(nick) {
        return this.users[Object.keys(this.users).find(id => this.users[id].nick === nick)]
    }

    getChannel(name) {
        return this.channels[name]
    }

    getUsersByChannel(channel, removeId) {
        removeId = removeId === undefined ? true : removeId

        let users = []
        for (let id in this.users) {
            if (this.getUser(id).channels[channel] !== undefined) {
                users.push(this.getUser(id))
            }
        }
        if (removeId) {
            return users.map(user => user.getPublicUser())
        }
        return users
    }

    getAvailableNick(nick) {
        let id = Object.keys(this.users).find(id => this.getUser(id).nick === nick)
        if (id === undefined) {
            return nick
        }
        let user = this.getUser(id)
        let suffix = 1
        do {
            if (Object.keys(this.users).find(id => this.getUser(id).nick === nick + '_' + suffix) === undefined) {
                break
            }
            suffix++
        } while (true)
        return nick + '_' + suffix
    }

    getIdlingUsers() {
        return Object.keys(this.users).filter(id => this.users[id].lastSeen < Date.now() - 5000 /* 5 seconds */ ).map(id => this.users[id])
    }

    addUser(user) {
        this.users[user.id] = user
        return this
    }

    addChannel(channel) {
        this.channels[channel.name] = channel
        return this
    }

    removeUser(id) {
        delete this.users[id]
        return this
    }

    removeChannel(name) {
        delete this.channels[name]
        return this
    }
}

module.exports = Store


app/models/User.js
"use strict"

const Channel = require('./Channel')

class User {
    constructor(nick) {
        this.nick = nick || "Anonymous"
        this.id = User.guid()
        this.channels = {}
        this.notices = []
        this.lastSeen = Date.now()
    }

    static guid() {
        function s4() {
            return Math.floor((1 + Math.random()) * 0x10000)
                .toString(16)
                .substring(1)
        }
        return s4() + s4() + '-' + s4() + '-' + s4() + '-' + s4() + '-' + s4() + s4() + s4()
    }

    static copy(other) {
        const user = new User(other.nick)
        user.id = other.id
        user.channels = {}
        for (let name in other.channels) {
            user.channels[name] = other.channels[name]
        }
        return user
    }

    isInChannel(channel) {
        return this.channels.indexOf(channel) >= 0
    }

    getPublicUser() {
        const publicUser = User.copy(this)
        publicUser.id = undefined
        publicUser.notices = []
        publicUser.lastSeen = undefined
        return publicUser
    }

    update() {
        this.lastSeen = Date.now()
    }
}

module.exports = User


server.js
const version = require('./package.json').version
const port = 8081

const express = require('express')
const cors = require('cors')
const favicon = require('serve-favicon')
const User = require('./app/models/User')
const Channel = require('./app/models/Channel')
const Store = require('./app/Store')

const bodyParser = require('body-parser')
const app = express()

app.use(favicon(__dirname + '/public/favicon.ico'))
app.use(bodyParser.urlencoded({
    extended: false
}))
app.use(bodyParser.json())
app.use(cors())
app.set('views', './public')
app.set('view engine', 'ejs')

let store = new Store()

// API

app.get('/', function(req, res) {
    res.render('api.ejs', {
        version: version
    })
})

// Info

app.get('/info/', function(req, res) {
    res.send({
        'version': version
    })
})

// List users

app.get('/users/', function(req, res) {
    res.send(Object.keys(store.users).map(id => store.users[id].getPublicUser()))
})

// Register

app.post('/users/register/:nick/', function(req, res) {
    const user = new User(store.getAvailableNick(req.params.nick))
    store.addUser(user)
    res.send(user)
})

// Disconnect

app.delete('/user/:id/disconnect/', function(req, res) {
    const user = store.getUser(req.params.id)
    if (user === undefined) {
        res.status(404).send({
            error: 'Unknown user ID'
        })
        return
    }
    Object.keys(user.channels).forEach(name => {
        notice({
            type: 'channelLeave',
            nick: user.nick,
            channel: store.getChannel(name)
        })
        const channel = store.getChannel(name)
        if (!channel.keep && store.getUsersByChannel(channel.name).length == 1) {
            store.removeChannel(channel.name)
        }
    })
    store.removeUser(user.id)
    res.send({
        status: 'Successfully disconnected from the server'
    })
})

// Whois

app.get('/users/whois/:nick/', function(req, res) {
    const id = Object.keys(store.users).find(id => store.users[id].nick === req.params.nick)
    const user = store.getUser(id)
    if (user === undefined) {
        res.status(404).send({
            'error': 'Unknown nick'
        })
    } else {
        res.send(user.getPublicUser())
    }
})

// List channels

app.get('/channels/', function(req, res) {
    res.send(Object.keys(store.channels).map(name => store.channels[name]))
})

// Join channel

app.put('/user/:id/channels/:channel/join/', function(req, res) {
    const user = store.getUser(req.params.id)
    if (user === undefined) {
        res.status(404).send({
            error: 'Unknown user ID'
        })
        return
    }
    let channel = store.getChannel(req.params.channel)
    if (channel === undefined) {
        channel = new Channel(req.params.channel, user.nick)
        store.addChannel(channel)
        notice({
            type: 'channelCreate',
            channel: channel
        })
    }
    if (channel.isBanned(user)) {
        res.status(404).send({
            error: 'Unable to join, you have been banned'
        })
        return
    }
    user.channels[channel.name] = channel
    notice({
        type: 'channelJoin',
        nick: user.nick,
        channel: channel
    })
    res.send({
        channel: channel,
        users: store.getUsersByChannel(channel.name, true)
    })
})

// Talk in channel

app.put('/user/:id/channels/:channel/say/', function(req, res) {
    const user = store.getUser(req.params.id)
    if (user === undefined) {
        res.status(404).send({
            error: 'Unknown user ID'
        })
        return
    }
    let channel = store.getChannel(req.params.channel)
    if (channel === undefined) {
        channel = new Channel(req.params.channel)
        store.addChannel(channel)
    }

    if (user.channels[channel.name] === undefined) {
        user.channels[channel.name] = channel
    }
    const message = req.body.message

    notice({
        type: 'channelMessage',
        nick: user.nick,
        channel: channel,
        message: message
    })
    res.send({
        status: 'Message sent correctly',
        message: message
    })
})

// Change channel description

app.put('/user/:id/channels/:channel/description/', function(req, res) {
    const user = store.getUser(req.params.id)
    if (user === undefined) {
        res.status(404).send({
            error: 'Unknown user ID'
        })
        return
    }
    const channel = store.getChannel(req.params.channel)
    if (channel === undefined) {
        res.status(404).send({
            error: 'Unknown channel'
        })
        return
    }
    const description = req.query.description
    if (description === undefined) {
        res.status(404).send({
            'error': 'Description cannot be null'
        })
    } else {
        channel.description = description

        notice({
            type: 'channelDescription',
            nick: user.nick,
            channel: channel
        })
        res.send({
            status: 'Changing the description',
            channel: channel
        })
    }
})

// Kick a user

app.delete('/user/:id/channels/:channel/kick/:nick/', function(req, res) {
    const user = store.getUser(req.params.id)
    if (user === undefined) {
        res.status(404).send({
            error: 'Unknown user ID'
        })
        return
    }
    const channel = store.getChannel(req.params.channel)
    if (channel === undefined) {
        res.status(404).send({
            error: 'Unknown channel'
        })
        return
    }
    const kickedUser = store.getUserByNick(req.params.nick)
    if (kickedUser === undefined) {
        res.status(404).send({
            error: 'Unknown username'
        })
        return
    }
    if (channel.owner !== user.nick) {
        res.status(404).send({
            error: 'You don\'t own the channel'
        })
        return
    }
    if (kickedUser.channels[channel.name] !== undefined) {
        delete kickedUser.channels[channel.name]
        if (channel !== undefined) {
            notice({
                type: 'userKick',
                nick: kickedUser.nick,
                channel: channel
            })
        }
        res.send({
            status: 'User successfully kicked',
            user: kickedUser,
            channel: channel
        })
    } else {
        res.send(404).send({
            error: "Not in channel, can't be kicked."
        })
    }
})

// Ban a user

app.post('/user/:id/channels/:channel/ban/:nick/', function(req, res) {
    const user = store.getUser(req.params.id)
    if (user === undefined) {
        res.status(404).send({
            error: 'Unknown user ID'
        })
        return
    }
    const channel = store.getChannel(req.params.channel)
    if (channel === undefined) {
        res.status(404).send({
            error: 'Unknown channel'
        })
        return
    }
    const bannedUser = store.getUserByNick(req.params.nick)
    if (bannedUser === undefined) {
        res.status(404).send({
            error: 'Unknown username'
        })
        return
    }

    channel.banned.push(bannedUser.nick)
    notice({
        type: 'userBan',
        nick: bannedUser.nick,
        channel: channel
    })
    res.send({
        status: 'User successfully banned',
        user: bannedUser,
        channel: channel
    })
})

// Unban a user

app.delete('/user/:id/channels/:channel/unban/:nick/', function(req, res) {
    const user = store.getUser(req.params.id)
    if (user === undefined) {
        res.status(404).send({
            error: 'Unknown user ID'
        })
        return
    }
    const channel = store.getChannel(req.params.channel)
    if (channel === undefined) {
        res.status(404).send({
            error: 'Unknown channel'
        })
        return
    }
    const bannedUser = store.getUserByNick(req.params.nick)
    if (bannedUser === undefined) {
        res.status(404).send({
            error: 'Unknown username'
        })
        return
    }

    const index = channel.banned.indexOf(bannedUser.nick)
    if (index > -1) {
        channel.banned.splice(index, 1)
    }
    notice({
        type: 'userUnban',
        nick: bannedUser.nick,
        channel: channel
    })
    res.send({
        status: 'User successfully unbanned',
        user: bannedUser,
        channel: channel
    })
})

// Fetch channel informations

app.get('/channels/info/:name/', function(req, res) {
    const name = req.params.name
    const channel = store.getChannel(name)

    if (channel === undefined) {
        res.status(404).send({
            'error': 'Unknown channel'
        })
    } else {
        res.send({
            channel: channel
        });
    }
})

// Keep channel

app.put('/user/:id/channels/:channel/keep', function(req, res) {
    const user = store.getUser(req.params.id)
    if (user === undefined) {
        res.status(404).send({
            error: 'Unknown user ID'
        })
        return
    }
    const channel = store.getChannel(req.params.channel)
    if (channel === undefined) {
        res.status(404).send({
            error: 'Unknown channel'
        })
        return
    }
    if (channel.owner !== user.nick) {
        res.status(404).send({
            error: 'You don\'t own the channel'
        })
        return
    }
    const keep = req.body.keep

    channel.keep = keep == true

    if (!channel.keep && store.getUsersByChannel(channel.name).length == 0) {
        store.removeChannel(channel.name)
    }

    notice({
        type: 'channelKeep',
        nick: user.nick,
        channel: channel
    })
    res.send({
        status: 'Changing the persistence',
        channel: channel
    })
})

// Give channel ownership

app.put('/user/:id/channels/:channel/owner/:nick/', function(req, res) {
    const user = store.getUser(req.params.id)
    if (user === undefined) {
        res.status(404).send({
            error: 'Unknown user ID'
        })
        return
    }
    const channel = store.getChannel(req.params.channel)
    if (channel === undefined) {
        res.status(404).send({
            error: 'Unknown channel'
        })
        return
    }
    const recipient = store.getUserByNick(req.params.nick)
    if (recipient === undefined) {
        res.status(404).send({
            error: 'Unknown username'
        })
        return
    }
    if (channel.owner !== user.nick) {
        res.status(404).send({
            error: 'You don\'t own the channel'
        })
        return
    }
    channel.owner = recipient.nick

    notice({
        type: 'channelOwner',
        channel: channel,
        nick: user.nick
    })
    res.send({
        status: 'Ownership transfered',
        channel: channel
    })
})

// Leave channel

app.delete('/user/:id/channels/:channel/leave/', function(req, res) {
    const user = store.getUser(req.params.id)
    if (user === undefined) {
        res.status(404).send({
            error: 'Unknow user ID'
        })
        return
    }
    const channel = store.getChannel(req.params.channel)
    if (user.channels[channel.name] !== undefined) {
        delete user.channels[channel.name]
        if (channel !== undefined) {
            if (!channel.keep && store.getUsersByChannel(channel.name).length == 0) {
                store.removeChannel(channel.name)
            }
            notice({
                type: 'channelLeave',
                nick: user.nick,
                channel: channel
            })
        }
    } else {
        res.send(404).send({
            error: "Not in channel, can't leave."
        })
    }
    res.send({
        status: 'Leaving the channel',
        channel: channel
    })
})

// Send private message

app.put('/user/:id/message/:nick/', function(req, res) {
    const sender = store.getUser(req.params.id)
    const recipient = store.getUserByNick(req.params.nick)
    const message = req.body.message
    if (sender === undefined) {
        res.status(404).send({
            error: 'Unknown user ID'
        })
        return
    }
    if (recipient === undefined) {
        res.status(404).send({
            error: 'Unknown username'
        })
        return
    }

    notice({
        type: 'privateMessage',
        sender: sender.nick,
        recipient: recipient.nick,
        message: message
    })
    res.send({
        status: 'Private message sent correctly',
        message: message,
        recipient: recipient
    })
})

// Fetch notices

app.get('/user/:id/notices/', function(req, res) {
    const user = store.getUser(req.params.id)
    if (user === undefined) {
        res.status(404).send({
            error: 'Unknown user ID'
        })
        return
    }
    user.update()
    const idlingUsers = store.getIdlingUsers()
    idlingUsers.forEach(user => {
        Object.keys(user.channels).forEach(name => {
            notice({
                type: 'channelLeave',
                nick: user.nick,
                channel: user.channels[name]
            })
            if (store.getUsersByChannel(name).length == 1) {
                store.removeChannel(name)
            }
        })
        store.removeUser(user.id)
    })
    res.send(user.notices)
    user.notices = []
})

// Notice

function notice(message) {
    message.time = new Date().toJSON()

    switch (message.type) {
        case 'channelJoin':
        case 'channelMessage':
        case 'channelLeave':
        case 'channelDescription':
        case 'channelKeep':
        case 'userKick':
        case 'userBan':
        case 'userUnban':
        case 'channelOwner':
            const channel = store.getChannel(message.channel.name)
            if (channel !== undefined) {
                const users = store.getUsersByChannel(channel.name, false)
                users.forEach(user => user.notices.push(message))
            }
            break

        case 'privateMessage':
            const recipient = store.getUserByNick(message.recipient)
            if (recipient !== undefined) {
                recipient.notices.push(message)
            }
            break

        case 'channelCreate':
            const users = store.users
            Object.keys(users).forEach(function(nick) {
                users[nick].notices.push(message)
            })
            break

        default:
    }
}

// Error handling

function error(req, res) {
    res.status(404).send({
        error: "Unknown route or method."
    })
    console.log('ERROR accessing ' + req.method + ' ' + req.path)
}

app.get('*', error)
app.post('*', error)
app.put('*', error)
app.delete('*', error)

// Server listening

if (__filename == process.argv[1]) {
    console.info('Diskuss v' + version + ' started on port ' + port)
    app.listen(port)
} else {
    module.exports = app
    module.exports.reset = function() {
        store = new Store()
    }
}
